File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* start, ListNode* end){
        ListNode* prev = NULL;
        while(start!=end){
            auto temp =  start->next;
            start->next = prev;
            prev = start;
            start = temp;
        }
        return prev;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(k==1) return head;
        int t;
        bool first = true;
        ListNode* ans;
        auto temp = head;
        ListNode* prevstart=NULL;
        while(true){
            auto start = temp;
            t=0;
            while(t<k){
                if(!temp){
                    return ans;
                }
                temp = temp->next;
                t++;
            }
            auto tempans = reverse(start,temp);
            if(first){
                ans = tempans;
                first = false;
            }
            if(prevstart){
                prevstart->next = tempans;
            }
            prevstart = start;
            start->next = temp;            
        }
        return ans;
    }
};

File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverse(self,head,end):
            prev = None
            temp = head
            while temp and temp!=end:
                nex = temp.next
                temp.next = prev
                prev = temp
                temp = nex
            return prev

    def reverseKGroup(self, head, k):
        r = head
        l = head
        cnt = 0
        ans = None
        prev = None
        prevStep = None
        while r:
            while r and cnt<k:
                prev = r
                r = r.next        
                cnt+=1
            if cnt!=k:
                if prevStep:
                    prevStep.next = l
                else:
                    return head
                # pass
                # return head
            else:
                newHead = self.reverse(l,r)
                if prevStep:
                    prevStep.next = newHead
                else:
                    ans = newHead
                prevStep = l
            l = r
            cnt = 0
            
        # l.next = self.reverseKGroup(r,k)        
        return ans